### ISSUE TITLE
command-not-found fails to suggest correct output to install

### ISSUE THREAD

**Issue by domenkozar** [NixOS/nixpkgs#56576]
```
$ dig
The program ‘dig’ is currently not installed. You can install it by typing:
  nix-env -iA nixos.bind

$ nix-shell -p bind --run "dig"
/tmp/nix-shell-31821-0/rc: line 1: dig: command not found
```

The problem is that dig binary is in `dnsutils` output.

Now, this can be fixed two ways:

# a) teach command-not-found about outputs

This would possibly need to extend narindex with information from what output are files coming from, so that it can suggest what output to use.

# b) restrict stdenv to contain `bin/` only in $out or $bin outputs

# Pros/Cons 

- (b) might mean we have to fix many derivations to comply
- if any of the packages need to exclude enforcement in (b) - for example due to closure size, they would again break command-not-found
- (b) does make things more consistent and predictable for humans

**Comment by domenkozar** [comment]
nix-index doesn't seem to have this bug as it indexes all files: https://github.com/bennofs/nix-index#a-files-database-for-nixpkgs

**Comment by vcunat** [comment]
(b) seems too restrictive, for example in this particular `dig` case.

**Comment by chreekat** [comment]
How many commands does this affect besides dig?

Why does bind have split outputs?

**Comment by danbst** [comment]
still not resolved

**Comment by wmertens** [comment]
I prefer `a)`, and since `nix-index` already handles it correctly and also ships a `command-not-found` replacement, I vote to make `nix-index` the default.

**Comment by wmertens** [comment]
This would be fixed by #39789

**Comment by Majiir** [comment]
Closing in favor of a more recently active duplicate: https://github.com/NixOS/nixpkgs/issues/129713
